### PR TITLE
Adjusted crop

### PR DESCRIPTION
**Pull request by AlexeyKozhevin**
Perform adjusting of shape to avoid padding after `rotate` and `scale`

### CODE REVIEW

File: seismiqb/src/crop_batch.py [hunk 1]
```diff
@@ -175,9 +186,13 @@ def make_locations(self, points, shape=None, direction=(0, 0, 0), eps=3,
...
             Batch with positions of crops in specified component.
         """
-        # pylint: disable=protected-access
+        # pylint: disable=too-many-arguments
 
         # Create all the points and shapes
+        crop_shape = shape
+        scale = scale if isinstance(scale, (list, tuple)) else [scale] * 3
+        shape = adjusted_shape_3d(shape, angle_1, angle_2, scale) if adjust else shape
+        print(shape)
```
Comment (AlexeyKozhevin) [inline seismiqb/src/crop_batch.py:195]:
```suggestion
```

File: seismiqb/src/crop_batch.py [hunk 2]
```diff
@@ -662,7 +679,7 @@ def _masks_to_horizons_post(self, horizons_lists, *args, dst=None, skip_merge=Fa
 
 
     @apply_parallel
-    def adaptive_reshape(self, crop, shape):
+    def adaptive_reshape(self, crop, shape=None):
```
Comment (SergeyTsimfer) [inline seismiqb/src/crop_batch.py:682]: Need to take the situation of different shapes for each crop into account

File: seismiqb/src/crop_batch.py [hunk 3]
```diff
@@ -671,6 +688,7 @@ def adaptive_reshape(self, crop, shape):
         shape : sequence
             Desired shape of resulting crops.
         """
+        shape = self.load_shape if shape is None else shape
```
Comment (SergeyTsimfer) [inline seismiqb/src/crop_batch.py:691]: Use `get` instead of `getattr`

File: seismiqb/src/utils.py [hunk 4]
```diff
@@ -455,6 +455,57 @@ def infer_tuple(value, default):
         value = tuple([item if item else default[i] for i, item in enumerate(value)])
     return value
 
+def adjusted_shape_2d(shape, angle):
```
Comment (SergeyTsimfer) [inline seismiqb/src/utils.py:458]: let's keep function names verb-like

File: seismiqb/src/utils.py [hunk 5]
```diff
@@ -455,6 +455,57 @@ def infer_tuple(value, default):
...
+        Target сrop shape.
+    angle : float
+
+    Returns
+    -------
+    tuple
+        Adjusted crop shape.
+    """
+    angle = np.abs(2 * np.pi * angle / 360)
+    limit = np.arctan(shape[1] / shape[0])
```
Comment (SergeyTsimfer) [inline seismiqb/src/utils.py:473]: Use `atan` from `math` module

File: seismiqb/src/utils.py [hunk 6]
```diff
@@ -455,6 +455,57 @@ def infer_tuple(value, default):
...
+    shape : tuple
+        Target сrop shape.
+    angle : float
+
+    Returns
+    -------
+    tuple
+        Adjusted crop shape.
+    """
+    angle = np.abs(2 * np.pi * angle / 360)
```
Comment (SergeyTsimfer) [inline seismiqb/src/utils.py:472]: use `abs` built-in

File: seismiqb/src/utils.py (hunk 5 above)
Comment (SergeyTsimfer) [reply seismiqb/src/utils.py:473]: Same for `cos`, `sin`, `ceil`

File: seismiqb/src/utils.py [hunk 7]
```diff
@@ -455,6 +455,57 @@ def infer_tuple(value, default):
...
+    limit = np.arctan(shape[1] / shape[0])
+    x_max, y_max = shape
+    if angle:
```
Comment (SergeyTsimfer) [inline seismiqb/src/utils.py:475]: Would not `angle != 0` be more explicit?

File: seismiqb/src/crop_batch.py [hunk 8]
```diff
@@ -912,16 +926,26 @@ def flip(self, crop, axis=0, seed=0.1, threshold=0.5):
         return crop
 
     @apply_parallel
-    def scale_2d(self, crop, scale):
+    def scale(self, crop, scale):
         """ Zoom in or zoom out along the first two axes of crop.
```
Comment (AlexeyKozhevin) [inline seismiqb/src/crop_batch.py:930]:
```suggestion
        """ Zoom in or zoom out along each axis of crop.
```

File: seismiqb/src/crop_batch.py [hunk 9]
```diff
@@ -1077,6 +1101,16 @@ def gaussian_filter(self, crop, axis=1, sigma=2, order=0):
         """ Apply a gaussian filter along specified axis. """
         return gaussian_filter1d(crop, sigma=sigma, axis=axis, order=order)
 
+    @apply_parallel
+    def central_crop(self, crop, shape):
+        """ Central crop of `self.crop_shape` shape. """
```
Comment (AlexeyKozhevin) [inline seismiqb/src/crop_batch.py:1106]:
```suggestion
        """ Central crop of defined shape. """
```

**Review by SergeyTsimfer — APPROVED** [review]

**Pull request merged**